### PR TITLE
fix/refactor bundle tests

### PR DESCRIPTION
**Pull request by ca-scribner**
# Summary
This PR fixes broken testing and refactors some tests for the bundle.

This change results in working CI that can be used on a locally deployed kubeflow (`tox -e tests -- --username XXX --password YYY`).  Note that this PR **does not** result result all problems in the github-driven integration tests as they have unrelated credential issues in AWS.

Other notes: I did not have a GPU available, so not sure if the `object_detection` pipeline currently works.  

Changes included:

### Changes to tests: 
* adds authentication flow (`tests/kf_authentication.py`) for pipeline tests (previously, auth was bypassed via port-forwarding in `run.sh`, but this stopped working)
* fixes selenium tests (mainly to adjust to minor webpage changes)
* fixes some pipeline tests
* improves the handling of dex credentials during testing, adding them as either command line arguments or pulled from environment variables during tests (previously some tests assumed default values or used hard coded ones)
* refactors `tests/test_kubectl.py` to check `juju status` of the bundle rather than look for specific pods or service accounts

### Changes to CI
* removes integration tests that were run on the github runner (these were run with the `kubeflow-edge` bundle, but the edge bundle is deprecated and no other bundle is small enough to run on github runners)
* pin AWS instance to a specific instance-type rather than resources to avoid over billing ([lp](https://bugs.launchpad.net/juju/+bug/1958574))
* set kubeflow/dex username/password as secrets to explicitly use them in the CI rather than rely on defaults

### Tests temporarily disabled
Some tests were temporarily disabled due to test failures.  These are documented in issues

* #431 `tests/test_selenium.py`/`test_notebook()` disabled due to upstream bug
* #432 `tests/test_pipeline.py`/`test_pipeline()`'s `mnist_pipeline` disabled as it was not working.  Not sure what is wrong
* #433 `tests/test_pipeline.py`/`test_pipeline()`'s `katib_pipeline` disabled as it was not working.  Not sure what is wrong

# Test instructions

(**see also the updated README.md file in the PR with more detailed instructions**)

* Deploy a local bundle-kubeflow (`juju deploy kubeflow --trust`)
* Patch istio (`kubectl patch role -n kubeflow istio-ingressgateway-operator -p '{"apiVersion":"rbac.authorization.k8s.io/v1","kind":"Role","metadata":{"name":"istio-ingressgateway-operator"},"rules":[{"apiGroups":["*"],"resources":["*"],"verbs":["*"]}]}'`)
* Set required config options
  ```
  juju config dex-auth static-username=user123@email.com
  juju config dex-auth static-password=user123
  juju config dex-auth public-url=http://10.64.140.43.nip.io
  juju config oidc-gatekeeper public-url=http://10.64.140.43.nip.io
  ```
* run tests (`tox -e tests -- -m lite --username user123@email.com --password user123`) (or `-m full`)

### CODE REVIEW

**Review by DomFleischmann — COMMENTED** [review]
This looks very promising, lets fix those lint errors and the `juju kubectl` issue fixed.

**Comment by DomFleischmann** [comment]
Just tried to run this on my own and I think you are missing instructions to set `KUBEFLOW_AUTH_USERNAME` and `KUBEFLOW_AUTH_PASSWORD`.

**Comment by ca-scribner** [comment]
Yeah adding the instructions to the readme is a great point.

Re the two env variables missing, I'm guessing you hit that from running `tox -e tests` and it breaking?  Did the error message there help?  Pytest should have given an error saying inputs were missing

**Comment by ca-scribner** [comment]
@DomFleischmann should be good to go now with updated instructions

File: README.md [hunk 1]
```diff
@@ -21,6 +21,34 @@ There are two possible paths, depending on your choice of Kubernetes:
...
+   sudo apt update
+   sudo pip3 install -r requirements.txt
+   sudo pip3 install tox
+```
+3. Run tests on your bundle with tox.  As many tests need authentication, make 
+   sure you pass the username and password you set in step (1) through 
+   environment variable or argument, for example:
+   1. full bundle (using command line arguments): 
+      ```
+      tox -e tests -- -m full --username user123@email.com --password user123`
```
Comment (DomFleischmann) [inline README.md:40]:
```suggestion
      tox -e tests -- -m full --username user123@email.com --password user123
```

File: README.md [hunk 2]
```diff
@@ -21,6 +21,34 @@ There are two possible paths, depending on your choice of Kubernetes:
 1. For any Kubernetes, follow the [installation instructions][install].
 1. On MicroK8s, you simply have to enable the [Kubeflow add-on][addon].
 
+## Testing
+
+To deploy this bundle and run tests locally, do the following:
+1. Set up Kubernetes, Juju, and deploy the bundle you're interested in (`kubeflow` or `kubeflow-lite`) using the [installation guide](https://charmed-kubeflow.io/docs/install/)
+2. Install test prerequisites:
+```bash
+   sudo snap install juju-wait --classic
```
Comment (DomFleischmann) [inline README.md:30]: These instructions are still missing  juju-kubectl.

File: README.md [hunk 3]
```diff
@@ -21,6 +21,34 @@ There are two possible paths, depending on your choice of Kubernetes:
...
+   sudo apt update
+   sudo pip3 install -r requirements.txt
+   sudo pip3 install tox
```
Comment (DomFleischmann) [inline README.md:33]: You are also missing this:
`sudo apt install -y firefox-geckodriver`

File: README.md [hunk 4]
```diff
@@ -21,6 +21,34 @@ There are two possible paths, depending on your choice of Kubernetes:
...
 
+## Testing
+
+To deploy this bundle and run tests locally, do the following:
+1. Set up Kubernetes, Juju, and deploy the bundle you're interested in (`kubeflow` or `kubeflow-lite`) using the [installation guide](https://charmed-kubeflow.io/docs/install/)
+2. Install test prerequisites:
+```bash
+   sudo snap install juju-wait --classic
+   sudo apt update
+   sudo pip3 install -r requirements.txt
```
Comment (DomFleischmann) [inline README.md:32]: Technically we are also missing installation instructions, for `pip3` although that could be skipped.

Charmcraft seems to be also needed in the tests.

**Review by DomFleischmann — CHANGES_REQUESTED** [review]
After the additional steps that are missing in the Instructions I was able to execute the tests successfully only the `cowsay_pipeline` failed with the following error:
```
tests/test_pipelines.py::test_pipelines[cowsay-cowsay_pipeline] 
--------------------------------------------------------------------------------------- live log call ----------------------------------------------------------------------------------------
INFO     root:_client.py:203 Failed to automatically set namespace.
Traceback (most recent call last):
  File "/home/ubuntu/bundle-kubeflow/.tox/tests/lib/python3.8/site-packages/kfp/_client.py", line 199, in __init__
    with open(Client.NAMESPACE_PATH, 'r') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/var/run/secrets/kubernetes.io/serviceaccount/namespace'
FAILED                                                                                                                                                                                 [ 75%]
tests/test_selenium.py::test_login 
```

File: README.md (hunk 2 above)
Comment (ca-scribner) [reply README.md:30]: I'm so confused about how this ran before.  Apparently copying and pasting from integration installs isn't enough haha

**Comment by ca-scribner** [comment]
@DomFleischmann ok I think this time I actually have all the dependencies

File: README.md [hunk 5]
```diff
@@ -21,6 +21,43 @@ There are two possible paths, depending on your choice of Kubernetes:
...
+   must include populating the `.kube/config` file with your Kubernetes cluster as the active
+   context
+1. Install test prerequisites:
+
+```bash
+   sudo snap install juju-wait --classic
+   sudo snap install juju-kubectl --classic
+   sudo snap install charmcraft --classic
+   sudo apt update
+   sudo apt install -y firefox-geckodriver
```
Comment (DomFleischmann) [inline README.md:39]: We should add `libssl-dev` here too.

Comment (DomFleischmann) [reply README.md:39]: After adding this dependency I was able to run the tests successfully @ca-scribner can you add it to the README? After that this should be good to merge

Comment (ca-scribner) [reply README.md:39]:
```suggestion
   sudo apt install -y libssl-dev firefox-geckodriver
```

**Review by DomFleischmann — APPROVED** [review]

**Pull request merged**